NNStreamer Filter:Tensorflow Error with TF 1.13

```
[   17s] [113/127] Compiling C++ object 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sta/tensor_filter_tensorflow_core.cc.o'.
[   17s] FAILED: ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sta/tensor_filter_tensorflow_core.cc.o 
[   17s] c++ -Iext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sta -Iext/nnstreamer/tensor_filter -I../ext/nnstreamer/tensor_filter -Igst/nnstreamer -I../gst/nnstreamer -I/usr/include/tensorflow -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/gstreamer-1.0 -I/usr/include/orc-0.4 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Werror -std=c++11 '-DVERSION="0.2.1"' -Wredundant-decls -Wwrite-strings -Wformat -Wformat-nonliteral -Wformat-security -Winit-self -Wmissing-include-dirs -Waddress -Wno-multichar -Wvla -Wpointer-arith -DHAVE_ORC=1 -DENABLE_TENSORFLOW=1 -DENABLE_TENSORFLOW_LITE=1 -D__TIZEN__=1 -O2 -g2 -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector -Wformat-security -fmessage-length=0 -Wl,--as-needed -feliminate-unused-debug-types --param=ssp-buffer-size=4 -fdiagnostics-color=never -m64 -march=nehalem -msse4.2 -mfpmath=sse -fasynchronous-unwind-tables -fno-omit-frame-pointer -g -std=gnu++11 -fPIC -pthread -Wno-sign-compare  -MD -MQ 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sta/tensor_filter_tensorflow_core.cc.o' -MF 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sta/tensor_filter_tensorflow_core.cc.o.d' -o 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sta/tensor_filter_tensorflow_core.cc.o' -c /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:422:9: error: 'void* TFBuffer::data() const' marked 'override', but does not override
[   17s]    void* data () const override { return data_; }
[   17s]          ^~~~
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc: In member function 'int TFCore::run(const GstTensorMemory*, GstTensorMemory*)':
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:471:19: error: use of deleted function 'TFBuffer::TFBuffer()'
[   17s]          buf = new TFBuffer;
[   17s]                    ^~~~~~~~
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:417:7: note: 'TFBuffer::TFBuffer()' is implicitly deleted because the default definition would be ill-formed:
[   17s]  class TFBuffer : public TensorBuffer {
[   17s]        ^~~~~~~~
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:417:7: error: no matching function for call to 'tensorflow::TensorBuffer::TensorBuffer()'
[   17s] In file included from /usr/include/tensorflow/tensorflow/core/framework/device_base.h:24:0,
[   17s]                  from /usr/include/tensorflow/tensorflow/core/common_runtime/device.h:39,
[   17s]                  from /usr/include/tensorflow/tensorflow/core/common_runtime/graph_runner.h:23,
[   17s]                  from /usr/include/tensorflow/tensorflow/core/common_runtime/shape_refiner.h:20,
[   17s]                  from /usr/include/tensorflow/tensorflow/c/c_api_internal.h:31,
[   17s]                  from /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.h:43,
[   17s]                  from /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:28:
[   17s] /usr/include/tensorflow/tensorflow/core/framework/tensor.h:645:12: note: candidate: tensorflow::TensorBuffer::TensorBuffer(void*)
[   17s]    explicit TensorBuffer(void* data_ptr) : data_(data_ptr) {}
[   17s]             ^~~~~~~~~~~~
[   17s] /usr/include/tensorflow/tensorflow/core/framework/tensor.h:645:12: note:   candidate expects 1 argument, 0 provided
[   17s] [114/127] Compiling C++ object 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sha/tensor_filter_tensorflow_core.cc.o'.
[   17s] FAILED: ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sha/tensor_filter_tensorflow_core.cc.o 
[   17s] c++ -Iext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sha -Iext/nnstreamer/tensor_filter -I../ext/nnstreamer/tensor_filter -Igst/nnstreamer -I../gst/nnstreamer -I/usr/include/tensorflow -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/gstreamer-1.0 -I/usr/include/orc-0.4 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Werror -std=c++11 '-DVERSION="0.2.1"' -Wredundant-decls -Wwrite-strings -Wformat -Wformat-nonliteral -Wformat-security -Winit-self -Wmissing-include-dirs -Waddress -Wno-multichar -Wvla -Wpointer-arith -DHAVE_ORC=1 -DENABLE_TENSORFLOW=1 -DENABLE_TENSORFLOW_LITE=1 -D__TIZEN__=1 -O2 -g2 -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector -Wformat-security -fmessage-length=0 -Wl,--as-needed -feliminate-unused-debug-types --param=ssp-buffer-size=4 -fdiagnostics-color=never -m64 -march=nehalem -msse4.2 -mfpmath=sse -fasynchronous-unwind-tables -fno-omit-frame-pointer -g -std=gnu++11 -fPIC -pthread -Wno-sign-compare  -MD -MQ 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sha/tensor_filter_tensorflow_core.cc.o' -MF 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sha/tensor_filter_tensorflow_core.cc.o.d' -o 'ext/nnstreamer/tensor_filter/584bea7@@nnstreamer_filter_tensorflow@sha/tensor_filter_tensorflow_core.cc.o' -c /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:422:9: error: 'void* TFBuffer::data() const' marked 'override', but does not override
[   17s]    void* data () const override { return data_; }
[   17s]          ^~~~
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc: In member function 'int TFCore::run(const GstTensorMemory*, GstTensorMemory*)':
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:471:19: error: use of deleted function 'TFBuffer::TFBuffer()'
[   17s]          buf = new TFBuffer;
[   17s]                    ^~~~~~~~
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:417:7: note: 'TFBuffer::TFBuffer()' is implicitly deleted because the default definition would be ill-formed:
[   17s]  class TFBuffer : public TensorBuffer {
[   17s]        ^~~~~~~~
[   17s] /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:417:7: error: no matching function for call to 'tensorflow::TensorBuffer::TensorBuffer()'
[   17s] In file included from /usr/include/tensorflow/tensorflow/core/framework/device_base.h:24:0,
[   17s]                  from /usr/include/tensorflow/tensorflow/core/common_runtime/device.h:39,
[   17s]                  from /usr/include/tensorflow/tensorflow/core/common_runtime/graph_runner.h:23,
[   17s]                  from /usr/include/tensorflow/tensorflow/core/common_runtime/shape_refiner.h:20,
[   17s]                  from /usr/include/tensorflow/tensorflow/c/c_api_internal.h:31,
[   17s]                  from /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.h:43,
[   17s]                  from /home/abuild/rpmbuild/BUILD/nnstreamer-0.2.1/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc:28:
[   17s] /usr/include/tensorflow/tensorflow/core/framework/tensor.h:645:12: note: candidate: tensorflow::TensorBuffer::TensorBuffer(void*)
[   17s]    explicit TensorBuffer(void* data_ptr) : data_(data_ptr) {}
[   17s]             ^~~~~~~~~~~~
[   17s] /usr/include/tensorflow/tensorflow/core/framework/tensor.h:645:12: note:   candidate expects 1 argument, 0 provided
[   17s] [115/127] Compiling C++ object 'tests/59830eb@@unittest_common@exe/common_unittest_common.cpp.o'.
[   18s] [116/127] Compiling C++ object 'tests/59830eb@@unittest_src_iio@exe/nnstreamer_source_unittest_src_iio.cpp.o'.
[   18s] [117/127] Compiling C++ object 'tests/tizen_capi/a680eff@@unittest_tizen_capi@exe/unittest_tizen_capi.cpp.o'.
[   21s] [118/127] Compiling C++ object 'tests/59830eb@@unittest_plugins@exe/nnstreamer_plugins_unittest_plugins.cpp.o'.
[   24s] [119/127] Compiling C++ object 'tests/59830eb@@unittest_sink@exe/nnstreamer_sink_unittest_sink.cpp.o'.
[   24s] ninja: build stopped: subcommand failed.
[   24s] error: Bad exit status from /var/tmp/rpm-tmp.f8x5yK (%build)
[   24s] 
[   24s] 
[   24s] RPM build errors:
[   24s]     Bad exit status from /var/tmp/rpm-tmp.f8x5yK (%build)
[   24s] 
[   24s] kohaku failed "build nnstreamer.spec" at Thu Jul 11 03:30:59 UTC 2019.
```
Result with 1.13.1-13.1 + absl headers



backward compatible solutions are recommended. (keep compatible with TF 1.09)

--- a/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc
+++ b/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.cc
@@ -419,7 +419,9 @@ class TFBuffer : public TensorBuffer {
   void* data_;
   size_t len_;
 
+#if (TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION < 13)
   void* data () const override { return data_; }
+#endif
   size_t size () const override { return len_; }
   TensorBuffer* root_buffer () override { return this; }
   void FillAllocationDescription (AllocationDescription* proto) const override {

--- a/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.h
+++ b/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_core.h
@@ -42,6 +42,7 @@
 #include <tensorflow/c/c_api.h>
 #include <tensorflow/c/c_api_internal.h>
 #include <tensorflow/core/public/session.h>
+#include <tensorflow/core/public/version.h>
 #pragma GCC diagnostic pop
 
 using namespace tensorflow;

